src/pybind/mgr/cephadm/services/service_registry.py

```diff
@@ -0,0 +1,76 @@
+Cephadm services. It dynamically discovers, imports, and registers all service
+classes located in the same directory, ensuring modularity and ease of extension.
+
+Key Features:
+- Automatically discovers and imports all service modules during initialization.
+- Registers service classes using the `@service_registry_decorator`.
+- Manages the lifecycle of service instances, initialized via `init_services`.
+- Provides a singleton `service_registry` for global access.
+
+Usage:
+1. Define a service class with a `TYPE` attribute and use `@service_registry_decorator`.
+2. Call `service_registry.init_services(mgr)` to initialize all registered services.
+3. Access services using `service_registry.get_service(service_type)`.
+"""
+
+import os
+import logging
+from typing import Type, Dict, TYPE_CHECKING
+import importlib
+import pkgutil
+
+if TYPE_CHECKING:
+    from cephadm.module import CephadmOrchestrator
+    from .cephadmservice import CephadmService
+
+logger = logging.getLogger(__name__)
+
+
+class CephadmServiceRegistry:
+    """Centralized registry for Cephadm services."""
+
+    def __init__(self) -> None:
+        self._services: Dict[str, "CephadmService"] = {}  # Initialized service instances
+        self._registered_services: Dict[str, Type["CephadmService"]] = {}  # Mapping of service types to classes
+
+    def register_service(self, service_type: str, service_cls: Type["CephadmService"]) -> None:
+        """Registers a service class to the registry."""
+        if service_type in self._registered_services:
+            logger.warning(f"Service type '{service_type}' is already registered.")
+        logger.info(f"Registering service: {service_type}")
+        self._registered_services[service_type] = service_cls
+
+    def init_services(self, mgr: "CephadmOrchestrator") -> None:
+        """Initializes and stores service instances using the provided orchestrator manager."""
+        self._discover_services()
+        for service_type, service_cls in self._registered_services.items():
+            logger.info(f"Initializing service: {service_type}")
+            self._services[service_type] = service_cls(mgr)
+
+    def _discover_services(self) -> None:
+        """Dynamically discovers and imports all modules in the current directory."""
+        current_package = __name__.rsplit(".", 1)[0]  # Get the package name
+        package_path = os.path.dirname(__file__)  # Directory of this file
+        for _, module_name, _ in pkgutil.iter_modules([package_path]):
+            logger.info(f"Importing service module: {current_package}.{module_name}")
+            importlib.import_module(f"{current_package}.{module_name}")
+
+    def get_service(self, service_type: str) -> "CephadmService":
+        """Retrieves an initialized service instance by type."""
+        return self._services[service_type]
+
+
+def service_registry_decorator(cls: Type["CephadmService"]) -> Type["CephadmService"]:
```

Comment:
naming nitpick: we don't need to say that it's a decorator in the name IMO, perhaps `register_cephadm_service_class` or just `register_service`? I think those are both clearer about what the decorator is doing and less redundant.

Reply:
I finally went with `register_cephadm_service` (I think it's more explicit) I just removed the `_class` suffix as it's somehow redundant in this case (the docs make it clear we have to placed it above the class).